File: leetcode/top150/39.cpp
#include <vector>
using namespace std;

class Solution {
    vector<vector<int>> ans;
    vector<int> path;
    void dfs(vector<int>& candidates, int currSum, int target, int i) {
        if (currSum == target) {
            ans.emplace_back(path);
            return;
        }
        if (currSum < target) {
            for (int j = i; j < candidates.size(); ++j) {
                path.push_back(candidates[j]);
                dfs(candidates, currSum + candidates[j], target, j);
                path.pop_back();
            }
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        dfs(candidates, 0, target, 0);
        return ans;
    }
};

File: leetcode/hot100_python/39.py
from typing import *

class Solution:
    def backtrack(self, ans: List[List[int]], candidates: List[int], path: List[int], target: int, i: int):
        if target == 0:
            ans.append(path.copy())
            return
        for j in range(i, len(candidates)):
            if candidates[j] > target:
                return
            path.append(candidates[j])
            self.backtrack(ans, candidates, path, target - candidates[j], j)
            path.pop()

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans, path = [], []
        candidates.sort()
        self.backtrack(ans, candidates, path, target, 0)
        return ans